### PR TITLE
Declare repository dependencies in child WORKSPACE files

### PR DESCRIPTION
**Pull request by kchodorow**
Bazel is moving to a model where child repository WORKSPACE files will be
parsed (see https://github.com/bazelbuild/bazel/issues/1943). This means
that top-level WORKSPACE files won't have to declare every dependency
(yay!) but, as a part of this, child repositories will have to start
declaring the repositories they use.

This should be a no-op, though, since the top-level WORKSPACE file
overrides all of these repository definitions.  So it doesn't really
matter what they are (hence the dummy paths).

### CODE REVIEW

**Comment by pmbethe09** [comment]
So for every new_go_repository, you have to know what repos it depends on?

**Comment by kchodorow** [comment]
Yes. Don't you have to now, anyway?  They repository definitions just go in the main WORKSPACE file.

**Comment by pmbethe09** [comment]
maintaining this 2-way list manually sounds hard

**Comment by pmbethe09** [comment]
<shrug>LGTM</shrug>

**Comment by kchodorow** [comment]
How are people doing this now? Just building, seeing what deps are missing, adding a new_go_repo, and rebuilding?

**Comment by pmbethe09** [comment]
pretty much.  there is a bare-bones tool in this repo 'wtool', which I use to add a new_go_repository, then build, then call wtool again for whatever was missing, etc.

**Comment by jayconrod** [comment]
We're thinking of giving Gazelle more capabilities for managing the the WORKSPACE file. See #388 and #389. Even if it doesn't end up writing new dependencies to WORKSPACE, it at least needs to be able to read them. If they're hidden in other WORKSPACE files, I don't know how we'll do that. Bazel doesn't really provide any capabilities for querying external repositories, as far as I know.

Personally, I'd much rather have a flat WORKSPACE file.

File: go/private/go_repository.bzl [hunk 1]
```diff
@@ -57,6 +57,13 @@ def _new_go_repository_impl(ctx):
   if result.return_code:
     fail("failed to generate BUILD files for %s: %s" % (
         ctx.attr.importpath, result.stderr))
+  ctx.file('WORKSPACE', """
+# Automatically generated by //go/private:go_repository.bzl
+workspace(name = "{workspace_name}")
+{repo_deps}
+""".format(
+    workspace_name = ctx.name,
+    repo_deps="\n".join(["local_repository(name='%s', path='/dummy')" % r for r in ctx.attr.repo_deps])))
```
Comment (jayconrod) [inline go/private/go_repository.bzl]: Using `local_repository` here doesn't make sense to me. Shouldn't Bazel report a conflict if you declare a `local_repository` in one place and, for example, a `git_repository` in another place? Seems like that should count as different versions.

Maybe there should be a special `use_repository` rule?

**Comment by jayconrod** [comment]
When are these declarations going to be mandatory in Bazel? I think we're going to have to improve our tooling to support this.

**Comment by kchodorow** [comment]
If this is going to cause you guys a lot of issues, I might be able to get away with not having per-repo declarations as long as the main repo declares them.  Let me think about it and see if I can implement that tomorrow...

File: go/private/go_repository.bzl (hunk 1 above)
Comment (kchodorow) [inline go/private/go_repository.bzl]: I could add a built-in "dummy" repository rule, if that works better for you guys.

**Comment by kchodorow** [comment]
Seems like that will be do-able.

**Pull request closed without merging**